### PR TITLE
Fix cumulative strain time calculation in Flashlight skill

### PR DESCRIPTION
**Pull request by MBmasher**
Consider this map:
```
object # | 1  2     3         4
         |-----------------------
time     | 0  30    80        150
```

Let `osuCurrent` be object 4. Initially, `osuPrevious` is set as object 3. At this point, `cumulativeStrainTime` should be calculated to be `150 - 80 = 70` as that is the difference between object 4 and 3. However, it is actually calculated as the `strainTime` of object 3 which is the difference between object 3 and object 2: `80 - 30 = 50`. This incorrect calculation continues with object 2 and 1 and all further objects.

This PR adds a new object called `lastObj` which is initially set to `osuCurrent` and represents the object directly after `currentObj` in time when looping backwards through the previous hitobjects list. Thus in the same scenario as above, if object 4 is `osuCurrent` and the loop is currently on `currentObj` = object 3, then `lastObj` is object 4.
Then, in the `cumulativeStrainTime` calculation, instead of adding `osuPrevious.StrainTime` (object 3), it adds the correct value which is `lastObj.StrainTime` (object 4).

All values get nerfed from this change, however they should be rebalanced after merging the other FL PRs: https://github.com/ppy/osu/pull/15665 https://github.com/ppy/osu/pull/15728

### CODE REVIEW

**Comment by smoogipoo** [comment]
[Spreadsheet](https://docs.google.com/spreadsheets/d/1m2AiQi2yPPASrV1MRHfF40g9dIl-ClOcgSOo2_95uJM/edit#gid=0) of SR/PP changes.

**Comment by MBmasher** [comment]
> [Spreadsheet](https://docs.google.com/spreadsheets/d/1m2AiQi2yPPASrV1MRHfF40g9dIl-ClOcgSOo2_95uJM/edit#gid=0) of SR/PP changes.

Looks like the most nerfed plays are on maps with a lot of bursts/streams/rhythm changes, which is expected since those were given inflated values from the incorrect calculations

overall the PR does what it was meant to do and there are no broken values, hopefully values will look better after the other PRs are merged

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs [hunk 1]
```diff
@@ -45,11 +45,17 @@ private double strainValueOf(DifficultyHitObject current)
                 var osuPrevious = (OsuDifficultyHitObject)Previous[i];
                 var osuPreviousHitObject = (OsuHitObject)(osuPrevious.BaseObject);
 
+                OsuDifficultyHitObject osuLastPrevious;
```
Comment (stanriders) [inline osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs:48]: This naming is quite confusing. Maybe something like `osuAfterPrevious` or `osuNextPrevious`?
I'm not sure I like naming object we're currently iterating through `previous` either, but can't really come up with anything better.

**Review by stanriders — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs (hunk 1 above)
Comment (apollo-dw) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs:48]: Something like `osuLoopCurrent` and `osuLoopNext` would be more descriptive

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs:48]: `osuLoop-` sounds good.

Comment (MBmasher) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs]: Resolved.

**Review by stanriders — DISMISSED** [review]

**Comment by bdach** [comment]
Merge conflict resolution required here.

**Review by smoogipoo — DISMISSED** [review]

**Comment by smoogipoo** [comment]
@MBmasher I've applied some more refactoring, following on from some discussion in Discord and also updated the description. Please make sure I haven't butchered your fix.

**Review by peppy — APPROVED** [review]

**Comment by MBmasher** [comment]
> @MBmasher I've applied some more refactoring, following on from some discussion in Discord and also updated the description. Please make sure I haven't butchered your fix.

Thank you, the change works fine.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**